internal/iamauth/token.go

```diff
@@ -0,0 +1,335 @@
+package iamauth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/go-secure-stdlib/strutil"
+)
+
+const (
+	amzHeaderPrefix    = "X-Amz-"
+	defaultIAMEndpoint = "https://iam.amazonaws.com"
+	defaultSTSEndpoint = "https://sts.amazonaws.com"
+)
+
+var defaultAllowedSTSRequestHeaders = []string{
+	"X-Amz-Algorithm",
+	"X-Amz-Content-Sha256",
+	"X-Amz-Credential",
+	"X-Amz-Date",
+	"X-Amz-Security-Token",
+	"X-Amz-Signature",
+	"X-Amz-SignedHeaders",
+}
+
+// BearerToken is a login "token" for an IAM auth method. It is a signed
+// sts:GetCallerIdentity request in JSON format. Optionally, it can include a
+// signed embedded iam:GetRole or iam:GetUser request in the headers.
+type BearerToken struct {
+	config *Config
+
+	getCallerIdentityMethod string
+	getCallerIdentityURL    string
+	getCallerIdentityHeader http.Header
+	getCallerIdentityBody   string
+
+	getIAMEntityMethod string
+	getIAMEntityURL    string
+	getIAMEntityHeader http.Header
+	getIAMEntityBody   string
+
+	iamEntityType           IAMEntityType
+	parsedCallerIdentityURL *url.URL
+	parsedIAMEntityURL      *url.URL
+}
+
+var _ json.Unmarshaler = (*BearerToken)(nil)
+
+type IAMEntityType string
+
+const (
+	IAMEntityTypeNone IAMEntityType = ""
+	IAMEntityTypeRole               = "role"
+	IAMEntityTypeUser               = "user"
+)
+
+func NewBearerToken(loginToken string, config *Config) (*BearerToken, error) {
```

Comment:
Some of the Vault parsing and validation of the login string is reframed here as this BearerToken type.